### DISCUSSION TITLE
How to run egui under windows xp？

### DISCUSSION THREAD

**Discussion by xixioba** [emilk/egui#5172] [General]
I want to run egui in Windows XP (32bit), so I refer to the [rust9x](https://github.com/rust9x/rust/wiki#compiling-the-compiler-and-standard-library) and compile rust 1.76 under **msys2+mingw32** environment, and helloworld project works perfectly
However, when I tested the demo of egui, a problem occurred, `dwmapi.dll not found`, and after doing some related searches, it seems that **dwm** is a feature added to Windows Vista. 
![截屏2024-09-26 18 06 35](https://github.com/user-attachments/assets/f566f3ba-9716-45ba-ba3b-eb8a394a3db1)

I don't know how to bypass this problem, and any help is greatly appreciated!

**Comment by xixioba** [comment]
After some trying, stuck here, if anyone has experience know how to fix it, it would be greatly appreciated
`point_ftol2 could not be located int the dynamic link library msvcrt.dll`
![截屏2024-10-05 14 38 43](https://github.com/user-attachments/assets/4597ad2d-d557-4b17-b37e-4c145a76f290)

**Reply by YgorSouza** [reply]
Have you tried linking the CRT statically? Although as discussed [here](https://users.rust-lang.org/t/statically-link-to-msvc-on-windows/108952), it still links to some libraries dynamically because they are supposed to be in every version of Windows that Rust supports, and Rust dropped Windows XP 4 years ago, so I'm guessing this is going to be a bit of an uphill battle.

**Reply by xixioba** [reply]
I have linked it, I think it may not be a problem with rust, because I also tried [imgui example](https://github.com/ocornut/imgui/discussions/8039), they have the same problem
Maybe the msys2 mingw32 environment cannot support windows xp gui programs? It really confuses me

**Reply by xixioba** [reply]
The effort brought some progress, **MSYS2 Mingw32** can run [OpenGL2-based GUI programs on Windows XP](https://github.com/ocornut/imgui/discussions/8039#discussioncomment-10863646), so maybe the problem is back to egui

**Comment by xixioba** [comment]
After some attempts, I can't get any further. 
I have successfully tested [**imgui example_glfw_opengl2**](https://github.com/ocornut/imgui/discussions/8039), it only needs a few dependencies, which can be met under windows xp. 
```
demo@DESKTOP-8PCFGOB MINGW32 /c/imgui-master/examples/example_glfw_opengl2
$ make
g++ -std=c++11 -I../.. -I../../backends -g -Wall -Wformat -D _USING_V110_SDK71_ -Wl,--subsystem,windows -Wl,--subsystem,console -static -static-libgcc -static-libstdc++ `pkg-config --cflags glfw3` -c -o main.o main.cpp
g++ -std=c++11 -I../.. -I../../backends -g -Wall -Wformat -D _USING_V110_SDK71_ -Wl,--subsystem,windows -Wl,--subsystem,console -static -static-libgcc -static-libstdc++ `pkg-config --cflags glfw3` -c -o imgui.o ../../imgui.cpp
g++ -std=c++11 -I../.. -I../../backends -g -Wall -Wformat -D _USING_V110_SDK71_ -Wl,--subsystem,windows -Wl,--subsystem,console -static -static-libgcc -static-libstdc++ `pkg-config --cflags glfw3` -c -o imgui_demo.o ../../imgui_demo.cpp
g++ -std=c++11 -I../.. -I../../backends -g -Wall -Wformat -D _USING_V110_SDK71_ -Wl,--subsystem,windows -Wl,--subsystem,console -static -static-libgcc -static-libstdc++ `pkg-config --cflags glfw3` -c -o imgui_draw.o ../../imgui_draw.cpp
g++ -std=c++11 -I../.. -I../../backends -g -Wall -Wformat -D _USING_V110_SDK71_ -Wl,--subsystem,windows -Wl,--subsystem,console -static -static-libgcc -static-libstdc++ `pkg-config --cflags glfw3` -c -o imgui_tables.o ../../imgui_tables.cpp
g++ -std=c++11 -I../.. -I../../backends -g -Wall -Wformat -D _USING_V110_SDK71_ -Wl,--subsystem,windows -Wl,--subsystem,console -static -static-libgcc -static-libstdc++ `pkg-config --cflags glfw3` -c -o imgui_widgets.o ../../imgui_widgets.cpp
g++ -std=c++11 -I../.. -I../../backends -g -Wall -Wformat -D _USING_V110_SDK71_ -Wl,--subsystem,windows -Wl,--subsystem,console -static -static-libgcc -static-libstdc++ `pkg-config --cflags glfw3` -c -o imgui_impl_glfw.o ../../backends/imgui_impl_glfw.cpp
g++ -std=c++11 -I../.. -I../../backends -g -Wall -Wformat -D _USING_V110_SDK71_ -Wl,--subsystem,windows -Wl,--subsystem,console -static -static-libgcc -static-libstdc++ `pkg-config --cflags glfw3` -c -o imgui_impl_opengl2.o ../../backends/imgui_impl_opengl2.cpp
g++ -o example_glfw_opengl2 main.o imgui.o imgui_demo.o imgui_draw.o imgui_tables.o imgui_widgets.o imgui_impl_glfw.o imgui_impl_opengl2.o -std=c++11 -I../.. -I../../backends -g -Wall -Wformat -D _USING_V110_SDK71_ -Wl,--subsystem,windows -Wl,--subsystem,console -static -static-libgcc -static-libstdc++ `pkg-config --cflags glfw3` -lglfw3 -lgdi32 -lopengl32 -limm32
Build complete for MinGW
```
However, **egui** seems to rely on many libraries, which can't be met under windows xp. 
```
"-Wl,-Bdynamic" "-lopengl32" "-lgdi32" "-luser32" "-lshell32" "-lwinapi_advapi32" "-lwinapi_cfgmgr32" "-lwinapi_gdi32" "-lwinapi_kernel32" "-lwinapi_msimg32" "-lwinapi_opengl32" "-lwinapi_shell32" "-lwinapi_shlwapi" "-lwinapi_synchronization" "-lwinapi_user32" "-lwinapi_winspool" "-lbcrypt" "-ladvapi32" "-lwindows.0.52.0" "-lkernel32" "-ladvapi32" "-lbcrypt" "-lkernel32" "-lntdll" "-luserenv" "-lws2_32" "-lkernel32" "-lntdll" "-lgcc_eh" "-l:libpthread.a" "-lmsvcrt" "-lmingwex" "-lmingw32" "-lgcc" "-lmsvcrt" "-luser32" "-lkernel32" "-Wl,--nxcompat" "-L" "D:\\gitlabs\\dist\\rust9x\\lib\\rustlib\\i686-pc-windows-gnu\\lib" "-o" "C:\\Users\\demo\\Downloads\\egui_glfw_gl-master\\egui_glfw_gl-master\\target\\i686-pc-windows-gnu\\release\\deps\\egui_glfw_gl-8adfcef6eac81fe2.exe" "-Wl,--gc-sections" "-no-pie" "-Wl,-O1" "-nodefaultlibs" "-D _USING_V110_SDK71_" "-Wl,--subsystem,console,5.01" "-Wl,--subsystem,windows"
```

Is there any chance that it can be like imgui?

**Reply by YgorSouza** [reply]
Egui itself doesn't depend on most of that, it's eframe that does. And the demo app also adds a bunch more dependencies. So maybe you can try a simpler example, with some of eframe's features disabled, or an alternative integration from the list in the readme <https://github.com/emilk/egui?tab=readme-ov-file#3rd-party-integrations>. Or as a last resort, write your own integration following the instructions that are in the readme as well, so you can use a backend that supports Windows XP.
